keps/sig-windows/2258-node-service-log-viewer/README.md

```diff
@@ -0,0 +1,373 @@
+Provide a cluster administrator with a streaming view of logs using kubectl
+without them having to implement a client side reader or logging into the node.
+This would work for:
+- Services on Linux worker and control plane nodes:
+  - That have systemd / journald support.
+  - That have services that log to `/var/log/`
+- Windows worker nodes (all supported variants) that log to `C:\var\log`,
+  System and Application logs, Windows Event Logs and Event Tracing (ETW).
+
+### Non-Goals
+- Providing support for non-systemd Linux distributions.
+- Reporting logs for nodes that have config or connection issues with the
+  cluster.
+- Getting logs from services that do not use /var/log/.
+
+## Proposal
+
+### Implement client for logs endpoint viewer (OS agnostic)
+- Extend `kubectl logs nodes` to work with node objects.
+- Implement a client for the `/var/log/` kubelet endpoint viewer.
+
+```
+# See what logs are available in masters in /var/log/
+kubectl logs nodes --role master --path=/
+
+# Display cron log file (/var/log/cron) from all masters
+kubectl logs nodes --role master --path=cron
+```
+
+### Linux distros with systemd / journald
+Supplement the the `/var/log/` endpoint viewer on the kubelet with a thin shim
+over the `journal` directory that shells out to journalctl. Then extend
+`kubectl logs` to also work with node objects.
+
+```
+# Show kubelet and crio journal logs from all masters
+kubectl logs nodes --role master --path journal -s kubelet -s crio
+```
+
+### Linux distributions without systemd / journald
+Running the new "kubectl logs nodes" command against services on nodes that do
+not use systemd / journald should return "OS not supported". However getting
+logs from `/var/log/` should work on all systems.
+
+### Windows
+Reuse the kubelet API for querying the Linux journal for invoking the
+`Get-WinEvent` cmdlet in a PowerShell.
+
+```
+# Show kubelet log file (/var/log/kubelet/kubelet.log) from all Windows worker nodes
+kubectl logs nodes --label kubernetes.io/os=windows --path kubelet/kubelet.log
+
+# Display docker runtime WinEvent log entries from a specific Windows worker node
+kubectl logs nodes <node-name> --service docker
+```
+
+### User Stories
+
+Consider a scenario where pods / containers are refusing to come up on certain
+nodes. As mentioned in the motivation section, troubleshooting this scenario
+involves the cluster administrator to SSH into nodes to scan the logs. Allowing
+them to use `kubectl logs` to do the same as they would to debug issues with a
+pod / container would greatly simply their debug workflow. This also opens up
+opportunities for tooling and simplifying automated log gathering. The feature
+can also be used to debug issues with Kubernetes services especially in Windows
+nodes that run as native Windows services and not as DaemonSets or Deployments.
+
+### Risks and Mitigations
+
+#### Large log files and events
+If the log that is attempted to be viewed is very large (GBs) there is
```

Comment:
Sorta think there should be a kubelet side throttle on how fast logs are streamed... just a though

Reply:
Good point. I will add that as something we should look into when graduating from alpha to beta after getting feedback from the field.

Reply:
I have added a few sentences about this and called out journald's rate limiting options.

Reply:
For Window should we consider allowing callers to specify -MaxEvents or set a cap limit how much data is streamed back?

Reply:
Looks like this is covered below with --tail :)